### ISSUE TITLE
smoothed cross entropy loss 

### ISSUE THREAD

**Issue by aa1607** [eladhoffer/utils.pytorch#4]
Hi , I noticed you changed cross_entropy.py lately to take account for ignore_index, but I think the new version might still be missing a step at the end? I was wondering in cross_entropy.py , when you calculate the smoothed cross entropy as the sum of the ce + kl, averaged over the number of tokens in the batch, shouldn't you subtract the padding tokens from the number of tokens in the denominator before you average?

```
if reduce:
kl = kl.mean() if size_average else kl.sum()

# for label smoothing with parameter eps:
if onehot_smoothing:
    entropy = -(math.log(1 - smooth_eps) + smooth_eps *
                math.log(smooth_eps / ((num_classes - 1) * (1 - smooth_eps))))
else:
    entropy = -(target * target.log()).sum()

if size_average:
    kl *= num_classes
    entropy /= logits.size(0)

```

here when you divide by logits.size(0), i think logits.size(0) = batch_size*sequence_length which includes padding_tokens in the total count. Shouldn't it be something like 

```
num_tokens = targets.ne(ignore_idx).sum()
...
kl = kl.sum() / num_tokens
...
entropy /= num_tokens 
```

Maybe with some epsilon introduced to make sure you don't divide by zero? Apologies if im wrong about this.
